pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -2208,6 +2279,46 @@ func (r *ReconcileClusterDeployment) setSyncSetFailedCondition(cd *hivev1.Cluste
 	return nil
 }
 
+// addOwnershipToTargetNamespace adds cluster deployment as an additional non-controlling owner to target namespace
+func (r *ReconcileClusterDeployment) addOwnershipToTargetNamespace(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger, name string) error {
+	cdLog = cdLog.WithField("namespace", name)
+
+	namespace := &corev1.Namespace{}
+	if err := r.Get(context.Background(), types.NamespacedName{Name: name}, namespace); err != nil {
+		cdLog.WithError(err).Error("failed to get namespace")
+		return err
+	}
+
+	annotationAdded := false
+	if namespace.Annotations[constants.CentralMachineManagementAnnotation] != cd.Name {
+		cdLog.Debug("Setting annotation on target namespace")
+		namespace.Annotations = k8sannotations.AddAnnotation(namespace.Annotations, constants.CentralMachineManagementAnnotation, cd.Name)
+		annotationAdded = true
+	}
+
+	cdRef := metav1.OwnerReference{
+		APIVersion:         cd.APIVersion,
+		Kind:               cd.Kind,
+		Name:               cd.Name,
+		UID:                cd.UID,
+		BlockOwnerDeletion: pointer.BoolPtr(true),
+	}
```

Comment:
"Cluster-scoped dependents can only specify cluster-scoped owners. In v1.20+, if a cluster-scoped dependent specifies a namespaced kind as an owner, it is treated as having an unresolveable owner reference, and is not able to be garbage collected." via https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents

I guess that means we're going to have to add a finalizer on the ClusterDeployment and only remove it after we've cleaned up the namespace?

Reply:
Now adding finalizer to ClusterDeployment.